### PR TITLE
image segmentation post

### PR DESCRIPTION
**Pull request by raymond-yuan**
Corresponding blog post can be found [here](https://docs.google.com/document/d/1y-ohUAMi2FjwgufFD0hudPt8A60TnHdysj4VcQrOxEA/edit#)

### CODE REVIEW

**Comment by lamberta** [comment]
Staged here in Colab: https://colab.research.google.com/github/tensorflow/models/blob/segmentation_blogpost/samples/outreach/blogs/segmentation_blogpost/5_Image_Segmentation.ipynb

I'm getting an error reading the train_masks.csv file. Is this meant to be a standalone example?

**Comment by raymond-yuan** [comment]
@lamberta You will need the `Carvana` dataset downloaded and moved to a /tmp/carvana folder  - I'm waiting to hear back from Josh (@random-forests)/licensing team to see if the we're able to use the dataset. 

This goes for the serving tutorial as well, we'll either need to host the data or write a script to get the data and process it (once we get permission from the source).

File: samples/outreach/blogs/segmentation_blogpost/5_Image_Segmentation.ipynb [hunk 1]
```diff
@@ -0,0 +1,1840 @@
...
...MT+/gX/75e5xwwu9Tq3qQAxy/5SUfsmPN9Eu9Xo+EpzQqlQqe5zE6Ogok1jwKusS0N/qsmDBpmoJkuvj/ToTLOsQ1GqG3VXyJB9/3G+tDqYYXT4vg0UEzEqbR9Y/VMbxZgbanYeRMY1mIMGmszGKxmLr8Qxrd3ntbJbzdEga7YV7gZ5aV5TIduWaLUKvVeObnP+HXjjyCZfuNYlurcREsAcsOr28eXzUe5NCo3MdrWrIjfDY8z8OmBlYxpXRDJ+aiDzMYDHNLX6+4IvIu4Hzg95VS07FD9wIniUhBRA4FDgd+BDwAHC4ih4pIHm2Id283ZeVyOWzbxnEcahWbQn6Cd7zj3YBPvjTFI/d9m23PlKPw+XH6jaqaNRiGKzMnKRQKFItFisUio6OjM9IeJAfjdgNz/O09K13cDimM+RLnySefjJZnSLteafYjvZJmJJw8Hq9v+D9NqxM3ru2m3Hi+yQ3aX7ukxiY8L3QzD7e4m3m45fN5nnviKTY+/gT+9A7y/jRHHXkMlgxjyXBT+a7rRpqXuB1NPp/Htu0mQTpp29NNDBoz/dTKXPZhBoNh7uj4qi0i/x14C7BSRDYBl6E9BQrAt4MO8z6l1J8qpR4XkW8CP0OrdE9VSnlBPqcB/4iOIf9VpdTjncpWCHVlQdiJ2xU8X4EdvB17JY75zbeipMKuPZt5+qnnwBvjmGOP1m+5hcDupeaC1XBBDgexsLNPqudDQaCTYGJZFtu2bWP16tXU6/UWA91uDHuT0x5J25hOGoekVieZf/Kz7/vYYVJxOe74Y9n28q7WMkKbEstvW0Yul8s0gk47J9J6BOU36tYsgKRphiJtF43lDbwO47UfOx7XkqngIsQ1NS3ti2mZwjoly43OVw4jMsWjjz7KqrWHMDExwZpD11AqlSLBw5NaNIfhh+dJcH29xjScb+lpsVzMHknRiOacvDZxTVscYySsmc8+zGAwzC2ykDu9I179GnXXV/4bkP5GrQfNYL/rNXnq+L7PP/793/C77/5DXIZQVPSAwR59rpeudUnTSiTLDd/Q07QF3Ux5JNO0m5rqRJbgFM8zjNcSCWu2T71ejxbRLBVH2+bfjYDXrm7t7Hna7U9O3SXrVK/XGcoXtMaiT2VEUqAJ74GItN6XvEVeCnraR7TA8+P7/i+VSoU3HP9b5HK55jXBEnOiTbZgQWwcN5huy4l
```
Comment (MarkDaoust) [inline samples/outreach/blogs/segmentation_blogpost/5_Image_Segmentation.ipynb]: honest question.. where does this come from? 
It feels funny, adding a probability to a log-probability... does anyone do `-log(dice_coef)`

Comment (MarkDaoust) [inline samples/outreach/blogs/segmentation_blogpost/5_Image_Segmentation.ipynb]: Any idea what the big jump in the middle of training is?

If you use `dice_loss` instead of `dice_coef` as the metric here, and use the same scale as on the "loss" panel, it makes things more comparable. You can directly the component of the total loss contributed the dice.

Comment (MarkDaoust) [inline samples/outreach/blogs/segmentation_blogpost/5_Image_Segmentation.ipynb]: The dice loss is [almost] just a differentiable IOU... right?

File: samples/outreach/blogs/segmentation_blogpost/5_Image_Segmentation.ipynb [hunk 2]
```diff
@@ -0,0 +1,1840 @@
...
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "colab_type": "text",
+    "id": "cl79rk4KKol8"
+   },
+   "source": [
+    "# Finding visual objects in images - Image Segmentation with tf.keras \n",
+    "In this tutorial we will learn how to segment images. **Segmentation** is the process of generating pixel-wise segmentations giving the class of the object visible at each pixel. For example, we could be identifying the location and boundaries of people within an image or identifying cell nuclei from an image. Formally, image segmentation refers to the process of partitioning an image into a set of pixels that we desire to identify (our target) and the background. \n",
```
Comment (MarkDaoust) [inline samples/outreach/blogs/segmentation_blogpost/5_Image_Segmentation.ipynb]: Could you add the standard set of github/colab buttons up top?

Like here:

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/eager/python/examples/generative_examples/image_captioning_with_attention.ipynb

File: samples/outreach/blogs/segmentation_blogpost/5_Image_Segmentation.ipynb (hunk 1 above)
[the remark this answers is not in the document]
Comment (raymond-yuan) [reply samples/outreach/blogs/segmentation_blogpost/5_Image_Segmentation.ipynb]: Sorry, I'm not sure how to save images within tensorflow.org, is it through github?

[the remark this answers is not in the document]
Comment (MarkDaoust) [reply samples/outreach/blogs/segmentation_blogpost/5_Image_Segmentation.ipynb]: Internal so we don't bloat the repo history.
Make an internal cl to: /third_party/tensorflow/docs_src/images/
Please add a directory, there are too many loose files there already.

File: samples/outreach/blogs/segmentation_blogpost/5_Image_Segmentation.ipynb [hunk 3]
```diff
@@ -0,0 +1,1840 @@
...
+      "startup": false,
+      "wait_interval": 0
+     }
+    },
+    "colab_type": "code",
+    "id": "9ej-e6cqmRgd"
+   },
+   "outputs": [],
+   "source": [
+    "df_train = pd.read_csv('/tmp/carvana/train_masks.csv')\n",
```
[the remark this answers is not in the document]
Comment (raymond-yuan) [reply samples/outreach/blogs/segmentation_blogpost/5_Image_Segmentation.ipynb]: done

File: samples/outreach/blogs/segmentation_blogpost/5_Image_Segmentation.ipynb (hunk 1 above)
[the remark this answers is not in the document]
Comment (raymond-yuan) [reply samples/outreach/blogs/segmentation_blogpost/5_Image_Segmentation.ipynb]: In terms of operations it's three to stack, transform, and then unstack, so might just leave as is (two operations, transform each)? What do you think?

[the remark this answers is not in the document]
Comment (raymond-yuan) [reply samples/outreach/blogs/segmentation_blogpost/5_Image_Segmentation.ipynb]: Updated

[the remark this answers is not in the document]
Comment (raymond-yuan) [reply samples/outreach/blogs/segmentation_blogpost/5_Image_Segmentation.ipynb]: done

[the remark this answers is not in the document]
Comment (raymond-yuan) [reply samples/outreach/blogs/segmentation_blogpost/5_Image_Segmentation.ipynb]: As I understood it, this [`tf.metrics.mean_iou`](https://www.tensorflow.org/api_docs/python/tf/metrics/mean_iou) simply just calculates the mean IOU between the labels and the prediction, ie not specific to `Estimators` or data streams etc. We're only using the tf.metrics.mean_iou since it doesnt support performing the mean over  different thresholds, we use it to calculate it over each threshold. 

We could just implement IOU, wdyt?

[the remark this answers is not in the document]
Comment (raymond-yuan) [reply samples/outreach/blogs/segmentation_blogpost/5_Image_Segmentation.ipynb]: The levels are somewhat arbitrary, but are to perform the mean over more confident predictions, which effectively adds a weight equal to the true labels. We don't include 0 because we don't want to score the background. 

The 1.0 isn't included, the max threshold we test against is `0.95`

Comment (raymond-yuan) [reply samples/outreach/blogs/segmentation_blogpost/5_Image_Segmentation.ipynb]: Err, yeah it is a little bit strange, but people achieved better results with it from the competition, e.g. [3rd place soln](https://www.kaggle.com/c/carvana-image-masking-challenge/discussion/40199)

Comment (raymond-yuan) [reply samples/outreach/blogs/segmentation_blogpost/5_Image_Segmentation.ipynb]: Right, they are almost the same, but I think there are differences when quantifying the the differences in performance over the entire set of predictions. I think dice tends to measure something closer to average performance, while the iou score measures something closer to the worst case performance.

Honestly, it's probably clearer for the user if we just remove the iou metric, I initially included it to illustrate how to add different custom metrics and give examples of how to implement them.

[the remark this answers is not in the document]
Comment (MarkDaoust) [reply samples/outreach/blogs/segmentation_blogpost/5_Image_Segmentation.ipynb]: Yes. 

But the advantage of stacking is that then you don't need the custom implementations of those two ops (`shift_img` and `flip_img`).

Something like:

```
def augmentation(img1,img2):
  imgs = tf.stack([img1,img2])
  imgs = tf.random_crop(imgs)
  imgs = tf.image.random_flip_left_right(imgs)
  return imgs[...,:-1], imgs[...,-1:]
```
Could be less code.

> What do you think?

It work as is. So NBD.

**Comment by raymond-yuan** [comment]
Thanks for the review @MarkDaoust! Also I realize reviewing this is pretty hairy, would it be easier if I made a py file download and updated the ipynb accordingly>

File: samples/outreach/blogs/segmentation_blogpost/5_Image_Segmentation.ipynb (hunk 1 above)
[the remark this answers is not in the document]
Comment (raymond-yuan) [reply samples/outreach/blogs/segmentation_blogpost/5_Image_Segmentation.ipynb]: Yeah, I think that's also more readable, will updated

Comment (MarkDaoust) [reply samples/outreach/blogs/segmentation_blogpost/5_Image_Segmentation.ipynb]: Cool. Add a link pointing to that discussion page?

[the remark this answers is not in the document]
Comment (MarkDaoust) [reply samples/outreach/blogs/segmentation_blogpost/5_Image_Segmentation.ipynb]: > not specific to `Estimators`

No. But designed for estimators, that's why they're so unnatural to use here.

> We could just implement IOU, wdyt?

+1, It might be simpler to implement IOU than to explain _what is happening here_.

[the remark this answers is not in the document]
Comment (MarkDaoust) [reply samples/outreach/blogs/segmentation_blogpost/5_Image_Segmentation.ipynb]: Done.

[the remark this answers is not in the document]
Comment (raymond-yuan) [reply samples/outreach/blogs/segmentation_blogpost/5_Image_Segmentation.ipynb]: Actually, this ends up being more code because of special cases, e.g. we only want to apply `random_hue` to the image so we have to unstack and then apply hue, then restack; for random flipping, it does not do it on a batch, so we have to unstack again; going to leave as is for now

Comment (raymond-yuan) [reply samples/outreach/blogs/segmentation_blogpost/5_Image_Segmentation.ipynb]: done

[the remark this answers is not in the document]
Comment (raymond-yuan) [reply samples/outreach/blogs/segmentation_blogpost/5_Image_Segmentation.ipynb]: I just removed it for now, I think that makes it the most clear and simple.

**Comment by MarkDaoust** [comment]
Here are a few uncontroversial fixes.

I renamed the file to delete the spaces.

Have a quick look at my changes, and then I'll merge it.

**Comment by raymond-yuan** [comment]
@MarkDaoust everything looks good to me, if you think it's ready, go ahead and merge? Thanks so much!
